File: packages/adb/src/test/lib/merkle-tree/saveLoadTree.test.ts
import fs from 'fs/promises';
import * as crypto from 'crypto';
import { 
    IMerkleTree,
    SortNode,
    FileHash,
    addFile,
    saveTree,
    loadTree,
    loadTreeVersion,
    traverseTreeSync,
    createTree,
    CURRENT_DATABASE_VERSION
} from '../../../lib/merkle-tree';
import { FileStorage } from 'storage';

describe('Merkle Tree Save/Load', () => {
    const TEST_FILE_PATH = './test-tree-v2.bin';
    
    /**
     * Helper function to create a file hash with a given name and content
     */
    function createFileHash(fileName: string, content: string = fileName): FileHash {
        // Create a proper 32-byte SHA-256 hash
        const hash = crypto.createHash('sha256').update(content).digest();
        return {
            fileName,
            hash,
            length: content.length,
            lastModified: new Date(),
        };
    }

    /**
     * Helper function to build a tree with the given file names
     */
    function buildTree(fileNames: string[]): IMerkleTree<any>{
        let merkleTree = createTree("12345678-1234-5678-9abc-123456789abc");
        
        for (const fileName of fileNames) {
            const fileHash = createFileHash(fileName);
            merkleTree = addFile(merkleTree, fileHash);
        }

        if (!merkleTree) {
            throw new Error('Failed to build the tree');
        }
        
        return merkleTree;
    }

    /**
     * Helper function to compare two binary trees
     */
    function compareTrees(original: SortNode | undefined, loaded: SortNode | undefined) {
        if (!original && !loaded) return;
        if (!original || !loaded) throw new Error('Tree structure mismatch');
        
        expect(loaded.nodeCount).toEqual(original.nodeCount);
        expect(loaded.leafCount).toEqual(original.leafCount);
        
        compareTrees(original.left, loaded.left);
        compareTrees(original.right, loaded.right);
    }
    
    // Clean up test file after each test
    afterEach(async () => {
        try {
            await fs.access(TEST_FILE_PATH);
            await fs.unlink(TEST_FILE_PATH);
        } catch (error) {
            // File doesn't exist, nothing to do
        }
    });
    
    test('should save and load a small tree correctly', async () => {
        // Create a simple tree with two files
        const originalTree = buildTree(['A', 'B']);
        
        // Save the tree to a file using V2 format
        await saveTree(TEST_FILE_PATH, originalTree, new FileStorage(""));
        
        // Load the tree from the file using V2 format
        const loadedTree = (await loadTree(TEST_FILE_PATH, new FileStorage("")))!;
        
        // Verify basic tree properties
        expect(loadedTree.metadata.totalFiles).toBe(originalTree.metadata.totalFiles);
        expect(loadedTree.sort?.nodeCount || 0).toBe(originalTree.sort?.nodeCount || 0);
        
        // Check that we have the expected file count
        expect(loadedTree.metadata.totalFiles).toBe(2);
        
        // Find leaf nodes to verify data integrity using binary tree traversal
        const leafNodes: SortNode[] = [];
        traverseTreeSync(loadedTree.sort, (node) => {
            if (node.nodeCount === 1 && node.fileName) {
                leafNodes.push(node);
            }
            return true; // Continue traversal
        });
        expect(leafNodes.length).toBe(2);
        
        // They should have filenames 'A' and 'B' (in some order)
        const fileNames = leafNodes.map(node => node.fileName).sort();
        expect(fileNames).toEqual(['A', 'B']);
        
        // Check that leaf nodes have proper hash properties
        for (const node of leafNodes) {
            expect(node.hash).toBeInstanceOf(Buffer);
            expect(node.hash.length).toBe(32);
        }
        
        // Verify node counts and leaf counts are preserved using binary tree traversal
        compareTrees(originalTree.sort, loadedTree.sort);
    });
    
    test('should save and load a complex tree correctly', async () => {
        // Create a larger tree with multiple levels
        const fileNames = ['A', 'B', 'C', 'D', 'E', 'F', 'G'];
        const originalTree = buildTree(fileNames);
        
        // Save the tree to a file
        await saveTree(TEST_FILE_PATH, originalTree, new FileStorage(""));
        
        // Load the tree from the file
        const loadedTree = (await loadTree(TEST_FILE_PATH, new FileStorage("")))!;
        
        // Verify basic tree properties
        expect(loadedTree.metadata.totalFiles).toBe(fileNames.length);
        expect(loadedTree.sort?.nodeCount || 0).toBe(originalTree.sort?.nodeCount || 0);
        
        // Verify leaf nodes have correct file names using binary tree traversal
        const leafNodes: SortNode[] = [];
        traverseTreeSync(loadedTree.sort, (node) => {
            if (node.nodeCount === 1 && node.fileName) {
                leafNodes.push(node);
            }
            return true; // Continue traversal
        });
        expect(leafNodes.length).toBe(fileNames.length);
        
        const loadedFileNames = leafNodes.map(node => node.fileName).sort();
        expect(loadedFileNames).toEqual(fileNames.sort());
        
        // Verify the structure using the existing compareTrees function
        compareTrees(originalTree.sort, loadedTree.sort);
        
        // Verify hash integrity - the root hash should match
        expect(loadedTree.sort?.hash.toString('hex')).toBe(originalTree.sort?.hash.toString('hex'));
    });
       
    test('should handle trees with special characters in file names', async () => {
        const specialChars = 'file-with-special-chars-!@#$%^&*()_+.txt';
        const tree = buildTree([specialChars]);
        
        await saveTree(TEST_FILE_PATH, tree, new FileStorage(""));

        const loadedTree = (await loadTree(TEST_FILE_PATH, new FileStorage("")))!;
        
        // the tree structure should be exactly the same
        expect(loadedTree.sort?.nodeCount || 0).toBe(tree.sort?.nodeCount || 0);
        expect(loadedTree.sort?.fileName).toBe(specialChars);
        expect(loadedTree.sort?.hash.toString('hex')).toBe(tree.sort?.hash.toString('hex'));        
    });
    
    test('should handle large trees efficiently', async () => {
        // Create a large tree with 100 files
        const fileNames = Array.from({ length: 100 }, (_, i) => `file-${i}.txt`);
        const originalTree = buildTree(fileNames);
        
        // Save the tree
        await saveTree(TEST_FILE_PATH, originalTree, new FileStorage(""));
        
        // Load the tree
        const loadedTree = (await loadTree(TEST_FILE_PATH, new FileStorage("")))!;
        
        // Verify tree properties
        expect(loadedTree.metadata.totalFiles).toBe(100);
        expect(loadedTree.sort?.nodeCount || 0).toBe(originalTree.sort?.nodeCount || 0);
                
        // Check that the trees are structurally identical using the compareTrees function
        compareTrees(originalTree.sort, loadedTree.sort);
    });
});

describe('loadTreeVersion', () => {
    const TEST_VERSION_FILE_PATH = './test-tree-version.bin';
    const storage = new FileStorage("");
    
    // Clean up test file after each test
    afterEach(async () => {
        try {
            await fs.access(TEST_VERSION_FILE_PATH);
            await fs.unlink(TEST_VERSION_FILE_PATH);
        } catch (error) {
            // File doesn't exist, nothing to do
        }
    });

    test('should load version from a valid tree file', async () => {
        // Create a tree and save it
        const originalTree = createTree("12345678-1234-5678-9abc-123456789abc");
        await saveTree(TEST_VERSION_FILE_PATH, originalTree, storage);
        
        // Load just the version
        const version = await loadTreeVersion(TEST_VERSION_FILE_PATH, storage);
        
        expect(version).toBe(CURRENT_DATABASE_VERSION);
    });

    test('should throw error for non-existent file', async () => {
        await expect(loadTreeVersion('./non-existent-file.bin', storage)).rejects.toThrow();
    });

    test('should return undefined for empty file', async () => {
        // Create an empty file
        await fs.writeFile(TEST_VERSION_FILE_PATH, Buffer.alloc(0));
        
        const version = await loadTreeVersion(TEST_VERSION_FILE_PATH, storage);
        
        expect(version).toBeUndefined();
    });

    test('should return undefined for file with less than 4 bytes', async () => {
        // Create a file with only 2 bytes
        await fs.writeFile(TEST_VERSION_FILE_PATH, Buffer.from([0x01, 0x02]));
        
        const version = await loadTreeVersion(TEST_VERSION_FILE_PATH, storage);
        
        expect(version).toBeUndefined();
    });

    test('should correctly read version from file with exactly 4 bytes', async () => {
        // Create a file with exactly 4 bytes representing version 42
        const versionBuffer = Buffer.alloc(4);
        versionBuffer.writeUInt32LE(42, 0);
        await fs.writeFile(TEST_VERSION_FILE_PATH, versionBuffer);
        
        const version = await loadTreeVersion(TEST_VERSION_FILE_PATH, storage);
        
        expect(version).toBe(42);
    });

    test('should read version from large file without loading entire file', async () => {
        // Create a file with version followed by lots of data
        const versionBuffer = Buffer.alloc(4);
        versionBuffer.writeUInt32LE(123, 0);
        
        // Add 1MB of additional data after the version
        const largeData = Buffer.alloc(1024 * 1024, 0xFF);
        const completeFile = Buffer.concat([versionBuffer, largeData]);
        
        await fs.writeFile(TEST_VERSION_FILE_PATH, completeFile);
        
        const version = await loadTreeVersion(TEST_VERSION_FILE_PATH, storage);
        
        expect(version).toBe(123);
    });

    test('should handle different version values correctly', async () => {
        const testVersions = [0, 1, 2, 3, 255, 65535, 4294967295]; // Test edge cases
        
        for (const expectedVersion of testVersions) {
            // Create file with specific version
            const versionBuffer = Buffer.alloc(4);
            versionBuffer.writeUInt32LE(expectedVersion, 0);
            await fs.writeFile(TEST_VERSION_FILE_PATH, versionBuffer);
            
            const version = await loadTreeVersion(TEST_VERSION_FILE_PATH, storage);
            
            expect(version).toBe(expectedVersion);
            
            // Clean up for next iteration
            await fs.unlink(TEST_VERSION_FILE_PATH);
        }
    });

    test('should work correctly with large tree files', async () => {
        // Create a complex tree with many files to demonstrate the function works with large files
        const fileNames = Array.from({ length: 100 }, (_, i) => `file-${i.toString().padStart(3, '0')}.txt`);
        let originalTree = createTree("12345678-1234-5678-9abc-123456789abc");
        
        for (const fileName of fileNames) {
            const fileHash: FileHash = {
                fileName,
                hash: crypto.createHash('sha256').update(fileName).digest(),
                length: fileName.length,
                lastModified: new Date(),
            };
            originalTree = addFile(originalTree, fileHash);
        }
        
        await saveTree(TEST_VERSION_FILE_PATH, originalTree, storage);
        
        // Load version from the large tree file
        const version = await loadTreeVersion(TEST_VERSION_FILE_PATH, storage);
        
        // Also load the full tree to compare
        const fullTree = await loadTree(TEST_VERSION_FILE_PATH, storage);
        
        // Verify both methods return the same version
        expect(version).toBe(CURRENT_DATABASE_VERSION);
        expect(fullTree?.version).toBe(CURRENT_DATABASE_VERSION);
        expect(version).toBe(fullTree?.version);
        
        // Verify the tree has the expected number of files
        expect(fullTree?.metadata.totalFiles).toBe(100);
    });
});

File: packages/merkle-tree-zig/src/test/saveLoadTree.test.zig
//
// Tests for saveTree, loadTree and loadTreeVersion (port of src/test/saveLoadTree.test.ts).
// (Zig: the TypeScript tests use FileStorage on temporary files; these use the in-memory storage.)
//

const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_verify = @import("merkle-verify.zig");
const memory_storage = @import("memory-storage.zig");
const errors = @import("utils-zig").errors;
const bson = @import("serialization-zig").bson;
const merkle_tree = merkle_tree_zig.merkle_tree;
const IMerkleTree = merkle_tree.IMerkleTree;
const SortNode = merkle_tree.SortNode;
const MemoryStorage = memory_storage.MemoryStorage;

//
// The Io used by the tests.
//
const io = std.testing.io;

//
// The file the tests save to.
//
const TEST_FILE_PATH = "test-tree-v2.bin";

//
// Helper function to build a tree with the given file names
//
fn buildTree(allocator: std.mem.Allocator, fileNames: []const []const u8) !IMerkleTree {
    var merkleTree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    for (fileNames) |fileName| {
        merkleTree = try merkle_tree.addItem(allocator, &merkleTree, try merkle_verify.createSha256HashedItem(allocator, fileName, fileName, fileName.len));
    }
    merkleTree.dirty = false;
    merkleTree.merkle = try merkle_tree.buildMerkleTree(allocator, merkleTree.sort);
    return merkleTree;
}

//
// Helper function to compare two binary trees
//
fn compareTrees(original: ?*const SortNode, loaded: ?*const SortNode) !void {
    if (original == null and loaded == null) {
        return;
    }
    if (original == null or loaded == null) {
        return error.TreeStructureMismatch;
    }
    try std.testing.expectEqual(original.?.nodeCount, loaded.?.nodeCount);
    try std.testing.expectEqual(original.?.leafCount, loaded.?.leafCount);
    try std.testing.expectEqual(original.?.size, loaded.?.size);
    try std.testing.expectEqualStrings(original.?.minName, loaded.?.minName);
    try std.testing.expectEqual(original.?.lastModified, loaded.?.lastModified);
    try compareTrees(original.?.left, loaded.?.left);
    try compareTrees(original.?.right, loaded.?.right);
}

//
// Saves a tree and loads it back.
//
fn saveAndLoad(allocator: std.mem.Allocator, tree: *const IMerkleTree) !IMerkleTree {
    var storage = MemoryStorage.init(allocator);
    try merkle_tree.saveTree(allocator, io, TEST_FILE_PATH, tree, storage.asStorage(), "FTRE");
    return (try merkle_tree.loadTree(allocator, io, TEST_FILE_PATH, storage.asStorage(), "FTRE")).?;
}

test "should save and load a small tree correctly" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const originalTree = try buildTree(allocator, &.{ "A", "B" });
    const loadedTree = try saveAndLoad(allocator, &originalTree);

    try std.testing.expectEqual(originalTree.sort.?.leafCount, loadedTree.sort.?.leafCount);
    try std.testing.expectEqual(originalTree.sort.?.nodeCount, loadedTree.sort.?.nodeCount);
    try std.testing.expectEqual(@as(u32, 2), loadedTree.sort.?.leafCount);

    var leaves = merkle_tree.iterateLeaves(SortNode, allocator, loadedTree.sort);
    const leafA = (try leaves.next()).?;
    const leafB = (try leaves.next()).?;
    try std.testing.expectEqualStrings("A", leafA.name.?);
    try std.testing.expectEqualStrings("B", leafB.name.?);
    try std.testing.expectEqual(@as(usize, 32), leafA.contentHash.?.len);
    try std.testing.expectEqual(@as(usize, 32), leafB.contentHash.?.len);

    try compareTrees(originalTree.sort, loadedTree.sort);
}

test "should save and load a complex tree correctly" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const fileNames = [_][]const u8{ "A", "B", "C", "D", "E", "F", "G" };
    const originalTree = try buildTree(allocator, &fileNames);
    const loadedTree = try saveAndLoad(allocator, &originalTree);

    try std.testing.expectEqual(@as(u32, fileNames.len), loadedTree.sort.?.leafCount);
    try std.testing.expectEqual(originalTree.sort.?.nodeCount, loadedTree.sort.?.nodeCount);
    try compareTrees(originalTree.sort, loadedTree.sort);

    // Verify hash integrity - the root hash should match
    try std.testing.expectEqualSlices(u8, originalTree.merkle.?.hash, loadedTree.merkle.?.hash);
    try std.testing.expectEqual(originalTree.merkle.?.nodeCount, loadedTree.merkle.?.nodeCount);
}

test "should handle trees with special characters in file names" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const specialChars = "file-with-special-chars-!@#$%^&*()_+.txt";
    const tree = try buildTree(allocator, &.{specialChars});
    const loadedTree = try saveAndLoad(allocator, &tree);

    try std.testing.expectEqual(tree.sort.?.nodeCount, loadedTree.sort.?.nodeCount);
    try std.testing.expectEqualStrings(specialChars, loadedTree.sort.?.name.?);
    try std.testing.expectEqualSlices(u8, tree.sort.?.contentHash.?, loadedTree.sort.?.contentHash.?);
}

test "should handle large trees efficiently" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var fileNames: [100][]const u8 = undefined;
    for (&fileNames, 0..) |*fileName, index| {
        fileName.* = try std.fmt.allocPrint(allocator, "file-{d}.txt", .{index});
    }
    const originalTree = try buildTree(allocator, &fileNames);
    const loadedTree = try saveAndLoad(allocator, &originalTree);

    try std.testing.expectEqual(@as(u32, 100), loadedTree.sort.?.leafCount);
    try std.testing.expectEqual(originalTree.sort.?.nodeCount, loadedTree.sort.?.nodeCount);
    try compareTrees(originalTree.sort, loadedTree.sort);
}

test "saveTree throws for a dirty tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTree(allocator, &.{"A"});
    tree.dirty = true;
    var storage = MemoryStorage.init(allocator);
    try std.testing.expectError(error.Thrown, merkle_tree.saveTree(allocator, io, TEST_FILE_PATH, &tree, storage.asStorage(), "FTRE"));
    try std.testing.expectEqualStrings("Tree is dirty. Cannot save. Make sure to rebuild the tree before saving.", errors.lastErrorMessage());
}

test "saveTree throws for an id that is not a UUID" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const tree = merkle_tree.createTree("test-tree");
    var storage = MemoryStorage.init(allocator);
    try std.testing.expectError(error.Thrown, merkle_tree.saveTree(allocator, io, TEST_FILE_PATH, &tree, storage.asStorage(), "FTRE"));
    try std.testing.expectEqualStrings("Invalid UUID", errors.lastErrorMessage());
}

test "saveTree throws like Buffer.writeUInt32LE for a last modified date before 1970" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = try buildTree(allocator, &.{"A"});
    tree.sort.?.lastModified = -315619200000; // 1960-01-01T00:00:00.000Z
    var storage = MemoryStorage.init(allocator);
    try std.testing.expectError(error.Thrown, merkle_tree.saveTree(allocator, io, TEST_FILE_PATH, &tree, storage.asStorage(), "FTRE"));
    try std.testing.expectEqualStrings("The value of \"value\" is out of range. It must be >= 0 and <= 4294967295. Received -74", errors.lastErrorMessage());
}

test "saves and loads an empty tree and its database metadata" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var tree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    var metadata: bson.BsonDocument = .empty;
    try metadata.put(allocator, "filesImported", .{ .number = 3 });
    tree.databaseMetadata = metadata;

    const loadedTree = try saveAndLoad(allocator, &tree);
    try std.testing.expect(loadedTree.sort == null);
    try std.testing.expect(loadedTree.merkle == null);
    try std.testing.expectEqual(@as(u32, 6), loadedTree.version);
    try std.testing.expect(!loadedTree.dirty);
    try std.testing.expect(loadedTree.databaseMetadata.?.eql(metadata));
}

test "loadTree returns undefined for a missing file" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    try std.testing.expect((try merkle_tree.loadTree(allocator, io, "missing.dat", storage.asStorage(), "FTRE")) == null);
}

test "should load version from a valid tree file" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const originalTree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    var storage = MemoryStorage.init(allocator);
    try merkle_tree.saveTree(allocator, io, "test-tree-version.bin", &originalTree, storage.asStorage(), "FTRE");
    try std.testing.expectEqual(@as(?u32, merkle_tree.CURRENT_DATABASE_VERSION), merkle_tree.loadTreeVersion(allocator, io, "test-tree-version.bin", storage.asStorage()));
}

test "should return undefined for non-existent file" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    try std.testing.expectEqual(@as(?u32, null), merkle_tree.loadTreeVersion(allocator, io, "non-existent-file.bin", storage.asStorage()));
}

test "should return undefined for empty file" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    try storage.putFile("test-tree-version.bin", "");
    try std.testing.expectEqual(@as(?u32, null), merkle_tree.loadTreeVersion(allocator, io, "test-tree-version.bin", storage.asStorage()));
}

test "should return undefined for file with less than 4 bytes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    try storage.putFile("test-tree-version.bin", &.{ 0x01, 0x02 });
    try std.testing.expectEqual(@as(?u32, null), merkle_tree.loadTreeVersion(allocator, io, "test-tree-version.bin", storage.asStorage()));
}

test "should correctly read version from file with exactly 4 bytes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    var versionBuffer: [4]u8 = undefined;
    std.mem.writeInt(u32, &versionBuffer, 42, .little);
    try storage.putFile("test-tree-version.bin", &versionBuffer);
    try std.testing.expectEqual(@as(?u32, 42), merkle_tree.loadTreeVersion(allocator, io, "test-tree-version.bin", storage.asStorage()));
}

test "should read version from large file without loading entire file" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    const completeFile = try allocator.alloc(u8, 4 + 1024 * 1024);
    @memset(completeFile, 0xFF);
    std.mem.writeInt(u32, completeFile[0..4], 123, .little);
    try storage.putFile("test-tree-version.bin", completeFile);
    try std.testing.expectEqual(@as(?u32, 123), merkle_tree.loadTreeVersion(allocator, io, "test-tree-version.bin", storage.asStorage()));
}

test "should handle different version values correctly" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var storage = MemoryStorage.init(allocator);
    const testVersions = [_]u32{ 0, 1, 2, 3, 255, 65535, 4294967295 }; // Test edge cases
    for (testVersions) |expectedVersion| {
        var versionBuffer: [4]u8 = undefined;
        std.mem.writeInt(u32, &versionBuffer, expectedVersion, .little);
        try storage.putFile("test-tree-version.bin", &versionBuffer);
        try std.testing.expectEqual(@as(?u32, expectedVersion), merkle_tree.loadTreeVersion(allocator, io, "test-tree-version.bin", storage.asStorage()));
    }
}

test "should work correctly with large tree files" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var originalTree = merkle_tree.createTree(merkle_verify.TEST_TREE_ID);
    var index: usize = 0;
    while (index < 100) {
        const fileName = try std.fmt.allocPrint(allocator, "file-{d:0>3}.txt", .{index});
        originalTree = try merkle_tree.addItem(allocator, &originalTree, try merkle_verify.createSha256HashedItem(allocator, fileName, fileName, fileName.len));
        index += 1;
    }
    originalTree.dirty = false;
    originalTree.merkle = try merkle_tree.buildMerkleTree(allocator, originalTree.sort);

    var storage = MemoryStorage.init(allocator);
    try merkle_tree.saveTree(allocator, io, "test-tree-version.bin", &originalTree, storage.asStorage(), "FTRE");

    const version = merkle_tree.loadTreeVersion(allocator, io, "test-tree-version.bin", storage.asStorage());
    const fullTree = (try merkle_tree.loadTree(allocator, io, "test-tree-version.bin", storage.asStorage(), "FTRE")).?;

    try std.testing.expectEqual(@as(?u32, merkle_tree.CURRENT_DATABASE_VERSION), version);
    try std.testing.expectEqual(merkle_tree.CURRENT_DATABASE_VERSION, fullTree.version);
    try std.testing.expectEqual(@as(u32, 100), fullTree.sort.?.leafCount);
}
